File: 프로그래머스/5/49190. 방의 개수/방의 개수.py
def solution(arrows):
    dx = [0, 1, 1, 1, 0, -1, -1, -1]
    dy = [1, 1, 0, -1, -1, -1, 0, 1]
    x, y = 0, 0
    visited_nodes = set()
    visited_edges = set()
    visited_nodes.add((x, y))
    answer = 0
    for arrow in arrows:
        for _ in range(2):
            nx, ny = x + dx[arrow], y + dy[arrow]
            edge = (x, y, nx, ny)
            reverse_edge = (nx, ny, x, y)
            if edge not in visited_edges:
                if (nx, ny) in visited_nodes:
                    answer += 1
                visited_edges.add(edge)
                visited_edges.add(reverse_edge)
            else:
                if (nx, ny) not in visited_nodes:
                    visited_nodes.add((nx, ny))
            visited_nodes.add((nx, ny))
            x, y = nx, ny
    return answer

File: 프로그래머스/5/49190. 방의 개수/방의 개수.cpp
#include <vector>
#include <unordered_set>
#include <utility>
using namespace std;

struct pair_hash {
    size_t operator()(const pair<int,int>& p) const {
        return hash<int>{}(p.first) ^ (hash<int>{}(p.second) << 1);
    }
};

struct edge_hash {
    size_t operator()(const pair<pair<int,int>, pair<int,int>>& e) const {
        return pair_hash{}(e.first) ^ (pair_hash{}(e.second) << 1);
    }
};

int solution(vector<int> arrows) {
    int answer = 0;
    int dx[8] = {0, 1, 1, 1, 0, -1, -1, -1};
    int dy[8] = {1, 1, 0, -1, -1, -1, 0, 1};
    int x = 0, y = 0;
    unordered_set<pair<int,int>, pair_hash> visited_nodes;
    unordered_set<pair<pair<int,int>, pair<int,int>>, edge_hash> visited_edges;
    visited_nodes.insert({x, y});
    for(int arrow : arrows){
        for(int i = 0; i < 2; ++i){
            int nx = x + dx[arrow];
            int ny = y + dy[arrow];
            pair<int,int> curr_node = {x, y};
            pair<int,int> next_node = {nx, ny};
            pair<pair<int,int>, pair<int,int>> edge = {curr_node, next_node};
            pair<pair<int,int>, pair<int,int>> reverse_edge = {next_node, curr_node};
            if(visited_edges.find(edge) == visited_edges.end()){
                if(visited_nodes.find(next_node) != visited_nodes.end()){
                    answer++;
                }
                visited_edges.insert(edge);
                visited_edges.insert(reverse_edge);
            }
            visited_nodes.insert(next_node);
            x = nx;
            y = ny;
        }
    }
    return answer;
}
